File: strategy/typescript/glados/agent/attacker.ts
import * as debug from "base/debug";
import { FriendlyRobot } from "base/robot";
import { parameterizeClass } from "base/types";
import * as World from "base/world";

import { ApplyForMainattacker } from "glados/agent/attacker/applyformainattacker";
import { Default } from "glados/agent/attacker/default";
import { DribblingGuard } from "glados/agent/attacker/dribblingguard";
import { Duel } from "glados/agent/attacker/duel";
import { DuelAssistant } from "glados/agent/attacker/duelassistant";
import { Move } from "glados/agent/attacker/move";
import { PenaltyPassiveAttackerOffense } from "glados/agent/attacker/penaltypassiveattackeroffense";
import { PenaltyShootout } from "glados/agent/attacker/penaltyshootout";
import { RunObjective } from "glados/agent/attacker/runobjective";
import { SelectObjective } from "glados/agent/attacker/selectobjective";
import { Stop } from "glados/agent/attacker/stop";
import { Agent } from "glados/agent/base/agent";
import { CheckableConstructor } from "glados/agent/base/behavior";
import { BallEscort } from "glados/agent/shared/ballescort";
import { PenaltyPassiveDefense } from "glados/agent/shared/penaltypassivedefense";
import { RemoveExtraRobot } from "glados/agent/shared/removeextrarobot";
import { MessageType } from "glados/control/messaging";

export class Attacker extends Agent {
	public beOffensive: boolean = false;

	protected _run() {
		if (this._activeBehavior) {
			this._messaging.sendBroadcast(MessageType.attackerFlag);
			this._messaging.sendToTrainerRepeated(MessageType.groupApplication, { name: "moves" });
		}

		debug.set("pool rating", this.rateRobot());
	}

	public getBehaviors(): CheckableConstructor[] {
		return [
			ApplyForMainattacker,
			SelectObjective,
			Move,
			Stop,
			PenaltyShootout,
			PenaltyPassiveDefense,
			PenaltyPassiveAttackerOffense,
			RemoveExtraRobot,
			Duel,
			DuelAssistant,
			parameterizeClass(BallEscort, false),
			DribblingGuard,
			RunObjective,
			Default,
		];
	}

	public static takeRobot(robots: FriendlyRobot[]): FriendlyRobot | undefined {
		for (let robot of robots) {
			if (robot.isVisible) {
				return robot;
			}
		}
		return undefined;
	}

	public keepRobot(): boolean {
		return this._robot.isVisible && this._robot !== World.FriendlyKeeper && this._robot.userControl == undefined;
	}

	// worse rating if robot is farther away from opponent goal
	public rateRobot(): number {
		if (this._activeBehavior && this._activeBehavior.forceKeepingInPool()) {
			return Infinity;
		}
		if (this._messaging.receiveTrainer(MessageType.mainAttacker) === this._robot) {
			return 1;
		}
		if (this._messaging.receiveTrainer(MessageType.exchangeRobot) === this._robot) {
			return 0;
		}
		return -World.Geometry.OpponentGoal.distanceTo(this._robot.pos);
	}
}

File: strategy/lua/marvin/agent/attacker.lua
--[[***********************************************************************
*   Copyright 2026 Robotics Erlangen e.V.                                 *
*   http://www.robotics-erlangen.de/                                      *
*   info@robotics-erlangen.de                                             *
*                                                                         *
*   This program is free software: you can redistribute it and/or modify  *
*   it under the terms of the GNU General Public License as published by  *
*   the Free Software Foundation, either version 3 of the License, or     *
*   any later version.                                                    *
*                                                                         *
*   This program is distributed in the hope that it will be useful,       *
*   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
*   GNU General Public License for more details.                          *
*                                                                         *
*   You should have received a copy of the GNU General Public License     *
*   along with this program.  If not, see <http://www.gnu.org/licenses/>. *
*************************************************************************]]

local Base = require "agent/base/agent"
local Attacker = Class("Agent.Attacker", Base)

local World = require "../base/world"
local debug = require "../base/debug"

local ApplyForMainattacker = require "agent/attacker/applyformainattacker"
local Default = require "agent/attacker/default"
local Duel = require "agent/attacker/duel"
local DuelAssistant = require "agent/attacker/duelassistant"
local FreeKick = require "agent/attacker/freekick"
local Move = require "agent/attacker/move"
local PassTiming = require "agent/attacker/passtiming"
local Penalty = require "agent/attacker/penalty"
local PenaltyDefensive = require "agent/attacker/penaltydefensive"
local PenaltyPassive = require "agent/shared/penaltypassive"
local PenaltyShootout = require "agent/attacker/penaltyshootout"
local Shoot = require "agent/attacker/shoot"
local Stop = require "agent/attacker/stop"
local BallEscort = require "agent/shared/ballescort"
local DoubleTouchGuard = require "agent/attacker/doubletouchguard"
local RescueFromDefenseArea = require "agent/shared/rescuefromdefensearea"

Attacker._behaviors = {
	ApplyForMainattacker,
	RescueFromDefenseArea,
	Move,
	Stop,
	PenaltyShootout,
	PenaltyDefensive,
	PenaltyPassive,
	Penalty,
	FreeKick,
	DoubleTouchGuard,
	Duel,
	DuelAssistant,
	BallEscort,
	PassTiming,
	Shoot,
	Default
}

function Attacker:init(robot, messaging)
	Base.init(self, robot, messaging)
	self.beOffensive = false
end

function Attacker:_run()
	if self._activeBehavior then
		assert(self._activeBehavior._send, "behavior message interface changed")
		self._activeBehavior._send.attackerFlag("all")

		local groupApplication = { name = "moves", payload = {} }
		self._activeBehavior._send.groupApplication("trainer", groupApplication)
	end

	debug.set("pool rating", self:rateRobot())
end

function Attacker.takeRobot(robots)
	for _, robot in ipairs(robots) do
		if robot.isVisible then
			return robot
		end
	end
end

function Attacker:keepRobot()
	return self._robot.isVisible and self._robot ~= World.FriendlyKeeper and not self._robot.userControl
end

-- worse rating if robot is farther away from opponent goal
function Attacker:rateRobot()
	if self._activeBehavior and self._activeBehavior:forceKeepingInPool()  then
		return math.huge
	end
	if self._inbox.mainAttacker().trainer == self._robot then
		return 0
	end
	return -World.Geometry.OpponentGoal:distanceTo(self._robot.pos)
end

return Attacker
